File: 0312-burst-balloons/0312-burst-balloons.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        
        n = len(nums)
        
        dp = [[0] * n for _ in range(n)]
        
        for length in range(1, n + 1):
            for i in range(n - length + 1):
                j = i + length - 1
                
                for k in range(i, j + 1):
                    left_dp = dp[i][k-1] if k - 1 >= 0 else 0
                    right_dp = dp[k+1][j] if k + 1 < n else 0
                    left_num = nums[i-1] if i - 1 >= 0 else 1
                    right_num = nums[j+1] if j + 1 < n else 1
                    
                    dp[i][j] = max(
                        dp[i][j],
                        left_dp + left_num * nums[k] * right_num + right_dp
                    )
        
        return dp[0][n-1]

File: 0312-burst-balloons/0312-burst-balloons.kt
class Solution {
    fun maxCoins(nums: IntArray): Int {
        val n = nums.size
        val newNums = IntArray(n+2)
        newNums[0] = 1
        newNums[n + 1] = 1
        for (i in 1..n) {
            newNums[i] = nums[i-1]
        }
        val dp = Array(n+2){IntArray(n+2)}
        for (len in 2..n + 1) {
            for (i in 0 until n+2-len) {
                val j = i+len
                for (k in i+1 until j) {
                    dp[i][j] = Math.max(dp[i][j], dp[i][k] + dp[k][j] + newNums[i] * newNums[k] * newNums[j])
                }
            }
        }
        return dp[0][n + 1]
    }
}